modules/prediction/evaluator/vehicle/semantic_lstm_evaluator.cc

```diff
@@ -30,17 +29,132 @@
 namespace apollo {
 namespace prediction {
 
+using apollo::common::TrajectoryPoint;
+using apollo::common::math::Vec2d;
+
 SemanticLSTMEvaluator::SemanticLSTMEvaluator() : device_(torch::kCPU) {
   evaluator_type_ = ObstacleConf::SEMANTIC_LSTM_EVALUATOR;
   LoadModel();
 }
 
 void SemanticLSTMEvaluator::Clear() {}
 
-bool SemanticLSTMEvaluator::Evaluate(Obstacle* obstacle_ptr) { return true; }
+bool SemanticLSTMEvaluator::Evaluate(Obstacle* obstacle_ptr) {
+  omp_set_num_threads(1);
+
+  obstacle_ptr->SetEvaluatorType(evaluator_type_);
+
+  Clear();
+  CHECK_NOTNULL(obstacle_ptr);
+  int id = obstacle_ptr->id();
+  if (!obstacle_ptr->latest_feature().IsInitialized()) {
+    AERROR << "Obstacle [" << id << "] has no latest feature.";
+    return false;
+  }
+  Feature* latest_feature_ptr = obstacle_ptr->mutable_latest_feature();
+  CHECK_NOTNULL(latest_feature_ptr);
+
+  if (!FLAGS_enable_semantic_map) {
+    ADEBUG << "Not enable semantic map, exit semantic_lstm_evaluator.";
+    return false;
+  }
+  cv::Mat feature_map;
+  if (!SemanticMap::Instance()->GetMapById(id, &feature_map)) {
+    return false;
+  }
+  // Process the feature_map
+  cv::cvtColor(feature_map, feature_map, CV_BGR2RGB);
+  cv::Mat img_float;
+  feature_map.convertTo(img_float, CV_32F, 1.0 / 255);
+  torch::Tensor img_tensor = torch::from_blob(img_float.data, {1, 224, 224, 3});
+  img_tensor = img_tensor.permute({0, 3, 1, 2});
+  img_tensor[0][0] = img_tensor[0][0].sub(0.485).div(0.229);
+  img_tensor[0][1] = img_tensor[0][1].sub(0.456).div(0.224);
+  img_tensor[0][2] = img_tensor[0][2].sub(0.406).div(0.225);
+
+  // Extract features of pos_history
+  std::vector<std::pair<double, double>> pos_history(20, {0.0, 0.0});
+  if (!ExtractObstacleHistory(obstacle_ptr, &pos_history)) {
+    ADEBUG << "Obstacle [" << id << "] failed to extract obstacle history";
+    return false;
+  }
+  // Process obstacle_history
+  torch::Tensor obstacle_pos = torch::zeros({1, 20, 2});
+  torch::Tensor obstacle_pos_step = torch::zeros({1, 20, 2});
+  for (int i = 0; i < 20; ++i) {
+    obstacle_pos[0][19 - i][0] = pos_history[i].first;
+    obstacle_pos[0][19 - i][1] = pos_history[i].second;
+    if (i == 19 || (i > 0 && pos_history[i].first == 0.0)) {
+      break;
+    }
+    obstacle_pos_step[0][19 - i][0] =
+        pos_history[i].first - pos_history[i + 1].first;
+    obstacle_pos_step[0][19 - i][1] =
+        pos_history[i].second - pos_history[i + 1].second;
+  }
```

Comment:
Avoid magic numbers. Give `20` a name such as `N`, then use `N - 1` to replace `19`.